File: Monad/monadReader.py
from collections import namedtuple

'''
Config struct and functions
'''
Conf = namedtuple('Conf', ['numberOfElems', 'filtro', 'transform'])

def numberOfElems(c):
  return c.numberOfElems
def filtro(c):
  return c.filtro
def transform(c):
  return c.transform

'''
Reader Monad class
'''
class Reader():
  # Reader e a
  def __init__(self, fun = None):
    self.r = fun

  def run(self, e):
    return self.r(e)

  # (a -> b) -> Reader e a -> Reader e b
  def fmap(self, f):
    return Reader(lambda e: f(self.r(e)))

  def unit(self, x):
    return Reader(lambda e: x)

  # Reader e a -> (a -> Reader e b) -> Reader e b
  def bind(self, fab):
    def f(e):
      a = self.r(e)
      rb = fab(a)
      return rb.run(e)
    return Reader(f)
    
def composeMonad(f, g):
  def k(a):
    return f(a).bind(g)
  return k

def ask(e):
  return e

def askFor(f):
  return Reader(ask).fmap(f)

'''
Example of use
'''
def recupera(tipo, xs):
  if tipo == "filtra":
    return filtra(xs)
  else:
    return transforma(xs)

def filtra(xs):
  f = composeMonad( composeMonad(aplicaFiltro, pega),
                    aplicaMap
                  )
  return f(xs)
  
def transforma(xs):
  f = composeMonad(pega, aplicaMap)
  return f(xs)
    
def aplicaFiltro(xs):
  filtroXS = lambda f: filter(f, xs)
  return askFor(filtro).fmap(filtroXS)

def aplicaMap(xs):
  mapXS = lambda f: map(f, xs)
  return askFor(transform).fmap(mapXS)

def pega(xs):
  pegaXS = lambda n: list(xs)[:n]
  return askFor(numberOfElems).fmap(pegaXS)
  
def bindTo(p, f):
    return askFor(p).bind(lambda x: Reader().unit(f(x)))

def even(x):
  return x%2==0
def square(x):
  return x*x
    
myCF = Conf(10, even, square)
print( list(recupera("transforma", range(1, 100)).run(myCF)) )
print( list(recupera("filtra", range(1, 100)).run(myCF)) )

File: Monad/monadReader.hs
import Control.Monad
import Control.Monad.Reader

data Config = Conf { alg :: String
                   , thr :: Double
                   , it  :: Int
                   }

go :: Int -> [Double] -> [Double]
go _ []     = []
go 0 xs     = xs
go i (x:xs) = (i' * x) : go (i-1) xs
  where i' = fromIntegral i

filterLess thr xs = filter (<thr) xs

f2 :: Config -> [Double] -> [Double]
f2 cfg xs = f3 cfg $ filterLess (thr cfg) xs

f3 :: Config -> [Double] -> [Double]
f3 cfg xs = go (it cfg) xs

algorithm :: Config -> [Double] -> [Double]
algorithm cfg xs | alg cfg == "f2" = f2 cfg xs
                 | otherwise       = f3 cfg xs

type ConfReader = Reader Config [Double]

f3' :: [Double] -> Reader Config [Double]
f3' xs = do
  t <- fmap it ask
  return (go t xs)

f2' :: [Double] -> Reader Config [Double]
f2' xs = do
  t  <- fmap thr ask
  f3' (filterLess t xs)

algorithm' :: [Double] -> Reader Config [Double]
algorithm' xs = do
  alg <- fmap alg ask
  if alg == "f2" then f2' xs else f3' xs

f3'' :: [Double] -> Reader Config [Double]
f3'' xs = fmap it ask >>= runGo
  where runGo t = return (go t xs)

f2'' :: [Double] -> Reader Config [Double]
f2'' xs = fmap thr ask >>= gof3
  where gof3 t = f3'' (filterLess t xs)


algorithm'' :: [Double] -> Reader Config [Double]
algorithm'' xs = fmap alg ask >>= choice
  where choice "f2" = f2'' xs
        choice _    = f3'' xs
--  where choice a | a == "f2" = f2'' xs
--                 | otherwise = f3'' xs

main = do
  let xs = [1.0, 2.0 .. 10.0]
      c = Conf "f2" 2.5 5
  print (algorithm c xs)
  print (runReader (algorithm' xs) c)
  print (runReader (algorithm'' xs) c)

